yt_idefix/data_structures.py

```diff
@@ -315,26 +317,38 @@ def _parse_parameter_file(self):
         self._setup_geometry()
 
     def _setup_geometry(self) -> None:
-        from_file = self.parameters.get("geometry")
-        from_user = self._geometry_from_user
-        if from_file is None and from_user is None:
+        from_bin = self.parameters.get("geometry")
+        from_definitions = self.parameters["definitions"].get("geometry")
+        from_input = self._geometry_from_user
+
+        if len({from_definitions, from_bin} - {None}) > 1:
+            raise RuntimeError(
+                "Geometries from disk file and definitions header do not match, got\n"
+                f" - {from_bin!r} (from {self.parameter_filename})\n"
+                f" - {from_definitions!r} (from {self._definitions_header})"
+            )
+
+        from_disk = from_bin or from_definitions
+
+        if from_disk is None and from_input is None:
             raise ValueError(
-                "Geometry couldn't be parsed from file. "
+                "Geometry couldn't be parsed from disk. "
                 "The 'geometry' keyword argument must be specified."
             )
-        elif from_user is not None:
-            if from_file is not None and from_user != from_user:
+
+        if from_input is not None:
+            if len({from_disk, from_input} - {None}) > 1:
```

Comment:
```suggestion
            if from_disk and from_input:
```
Is it enough and more straightforward here?

Reply:
I don't think so. The more explicit way to implement my check is
```python
if from_disk is not None and from_input is not None and from_disk != from_input:
```
maybe I should use that instead if you think the current state is confusing. This is a case where having variable that can be more than one type (`str` or `None`) quickly escalates into nonesense... maybe I could just use `""` as default value instead of None 🤔